### ISSUE TITLE
remove schema restriction on texts around inlines

### ISSUE THREAD

**Issue by ianstormtaylor** [ianstormtaylor/slate#2865]
#### Do you want to request a _feature_ or report a _bug_?

Debt.

#### What's the current behavior?

Right now the core schema will always ensure that inline nodes are surrounded by text nodes. This was to ensure that when inserting text you can always insert on either side of the inline node. 

However it's a bit of an intense restriction, and results in the data model itself having extra "junk" in it.

#### What's the expected behavior?

Instead, I think we may be able to solve this with editor queries. With a query like...

```js
editor.resolveInsertionPoint(point)
```

It could ensure that a text node exists to type into if we're at the leading/trailing edge of an inline node. This would also be an override point for folks who want to implement "sticky" inlines.

That way the restriction can be editor-level instead of data-level, resulting in more flexibility for people with advanced use cases.

**Comment by justinweiss** [comment]
This is an interesting idea. Do you think you would have Slate delegate to this query for that last bit of cursor normalization (https://github.com/ianstormtaylor/slate/blob/master/packages/slate/src/models/point.js#L417, for example?)

**Comment by justinweiss** [comment]
> Slightly different than an editor.getInsertionPoint which you wouldn't expect to change the document at all.

Would you expect `getInsertionPoint` to be able to return points that don't refer to real nodes in the document (For example, if it returns a text node following an inline that we haven't inserted)? Or would it be possible for it to return a different point than `resolveInsertionPoint`?

**Comment by ianstormtaylor** [comment]
@justinweiss thanks for your questions!

In thinking about this more, I think it might be trickier than it's worth, because of the need to represent fake points like you mention. We'd have two choices:

1. Simplify the working logic, at the expense of normalize the data model.
2. Simplify the data model, at the expense of complicating the working logic.

Right now we go with **1.** because it makes working with Slate documents much easier. Although I'd always prefer to go with **2.** if possible because it keeps the data "purest". 

One option for **2.** is to introduce the concept of "fake" points in the document. But that seems pretty unwieldy because pretty much every place that deals with points would need to check if the point was real or not first.

Another option would be use `resolveInsertionPoint` like I mentioned. But then you have this implicit knowledge that has to occur that knows that even though the cursor is inside an `Inline` for instance, it won't actually insert there. And rendering becomes more complex because you probably wouldn't want to render the inline node as selected at that point? Not sure.

Not to mention that getting to true purity would also not require void nodes having text nodes at all. And that introduces even more inconsistency in the data that needs to then be accounted for in `n` places.

---

Not saying this is closed forever, but it does seem like it might be more complex than it's worth? What do you think?

I still think keeping the data model as pure and simple as possible is a good goal, just not sure if there's a way to do it without pushing that complexity into many many places.

**Comment by justinweiss** [comment]
Thanks for the thoughts! I agree that this is more complex than it’s worth. I would still love a plugin-overridable `getInsertionPoint`, to make things like slate-sticky-inlines easier to build. Being able to edit the boundaries of links and marked text is _such_ a huge improvement over other text editors. Maybe that belongs in a different issue?

I haven’t had too much trouble with the data model as-is, except when normalizing documents converted from HTML that don’t follow the Slate basic schema (empty nodes not surrounding inlines, sibling text nodes, etc.). That’s been easy enough to fix with a pure-JS-only custom pass after deserialization, before handing things off to Slate.

**Comment by ianstormtaylor** [comment]
I completely agree, I want the `getInsertionPoint` behavior to be overridable too. Another issue for it sounds good just to keep track. 

As for data model that makes sense. I think my use case is probably pretty far out there in terms of architecture, so it might not make as much sense. (Although I think that’s generally a good thing for making sure the framework is well designed!)

Can you give me an example of your deserialization troubles? Just curious to see if there are ways to ease them. 

Thanks!

**Comment by justinweiss** [comment]
Sure thing!

There were two major areas where our deserializer would create documents that needed to be normalized after-the-fact. 

When deserializing some HTML into an inline node, I think we ran into examples with where that inline node (a link, for example), didn't have surrounding DOM text nodes to get deserialized to surrounding text nodes. 

We also ran into some places where we were deserializing into runs of text nodes that had the same marks, so Slate would merge them into a single text node. 

Some of this we could fix during deserialization. Some was unavoidable, because our deserializers couldn't easily get a big-picture view of their node, and even if they could, it seems like that isn't really their job.

Some medium-sized real-world documents would create thousands or tens of thousands of operations to normalize those documents. At the time (I think this has changed), each operation contained a value, so the browser could run out of memory as it generated tens of thousands of copies of a large value. Even if it didn't, every time we needed to find nodes that needed to be normalized, it would walk the entire document, which could take quite a while as we kept generating new operations and doing more walks.

Since our use of deserializing will always create a brand-new document, we don't care _at all_ about calling Change methods and generating Operations during deserialization, we just want a mostly good document to start with. So before ever putting the deserialized objects through `Value.fromJSON`, we do a cleanup pass that walks the tree and does the same things as the default serializers (inserts text nodes around inlines, merges text nodes / leaves that can be merged).

I don't know how this is solvable inside Slate. We could port the default normalizers to operate on pure JS objects during `fromJSON`, but sometimes you _might_ want those normalizing operations to be generated when loading a document. (Maybe an option?)

Or maybe the initial normalizing could be done in a single pass (or a few passes), just generating a bunch of operations that all get transformed against previous normalizing operations, so you aren't walking the document again after each operation is applied, just after each large batch of operations. But the time taken by that would still grow based on both the size of the document and the number of operations generated, which is where the fundamental performance problem is. You avoid some tree walks that way, though.

I have seen this has come up in the Slack a few times. It's easy to solve for yourself once you know what's happening. But it's not that easy to discover, especially until you go live or do performance testing.

**Comment by ianstormtaylor** [comment]
Ah okay, I didn’t realize it was all inside the normalization logic where the problems are. Having the multi-pass solution for performance sounds like it could be promising. And once we’ve dropped Immutable that will make it a bit less costly too hopefully.
